### PR TITLE
Add PLEP 8 on switching to calendar versioning

### LINKED ISSUE #38: Amend PLEP 5 with updated versioning scheme
[PLEP 5](https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0005.rst) currently specifies that PlasmaPy uses [semantic versioning](https://semver.org/) (SemVer).  SemVer requires that (after the `1.0.0` release) any backward incompatible changes result in an incremented major version number. This is not so practical for a scientific software package that has a continual, "perpetual beta" development model such that there are likely to be some backward incompatible changes needed during every non-patch release.  (I believe there were some extensive discussions about this in SunPy.)

We haven't reached consensus, but some options and an incomplete tradeoff analysis.
 1. 0️⃣ Keep using `0.y.z` versioning for the time being
   - Pros: Emphasizes that we are still in the development phase.  Consistency with prior versions.  Consistency with packages like Astropy, NumPy, SciPy, matplotlib, xarray, etc.  We can procrastinate making this decision for longer than we have been 😬.
   - Cons: Keeping with this scheme for a long time will continue to imply package instability & immaturity.  Won't be able to tell how old a version is just by the version number.
 2. 📆 Switch the calendar based versioning (e.g., PyCharm has versions like `2021.1.2` where `1` refers to the release number in a given year) now or at some point later.
   - Pros: Lets people know how old their version of PlasmaPy is.  Consistency with software like dask, Ubuntu, & PyCharm.  Easy to understand just from reading it.  Would let users know in case there is any time-sensitive functionality (i.e. if a version is released before an experiment or diagnostic began operations).
   - Cons: Inconsistency with prior versions.  Inconsistency with most of the packages that we depend on.  Won't be able to tell how mature a package is just by the version number.  Would need to be a permanent change to preserve monotonicity.  
 3. 🔢 Switch to [a versioning scheme like SunPy's](https://github.com/sunpy/sunpy-SEP/blob/master/SEP-0009.md#sep-0009----release-pattern-for-sunpy-core) where the major version number is incremented for every long-term support release.
  - Pros: Consistency with prior version scheme.  Consistency with packages like Astropy, NumPy, SciPy, matplotlib, xarray, etc.  LTS releases are predictable and built into the schedule. Would be able to switch to calendar based versioning later on.
  - Cons: Won't be able to tell how old a version is just by the version number.  Won't be able to tell a release is LTS just from the version number without being familiar with how these versioning schemes usually work.

If there are any pros and cons that this is missing, please post below.

### PR DESCRIPTION
**Pull request by namurphy**
This PR follows up on a discussion in #38 and elsewhere in the plasmapythoniverse about switching from semantic versioning to date-based versioning.  There's a discussion of the tradeoffs in #38.  

I'm basing the scheme in the first draft on Ubuntu's, which has versions like ``22.04`` and ``22.10``, which were the releases from April and October 2022.  The year format is concise, and it's more clear with the zero-padded month that it actually represents a month.  (I got confused earlier today with pip ``v21.3``, since I thought that the minor version number meant the month when it actually didn't.)

This new PLEP would repeal PLEP 5, but since it's a complete rewrite, I decided on doing a new PLEP so as to make the history more apparent.

Closes #38.

### CODE REVIEW

File: PLEP-0008.rst [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+* Calendar-based version numbers of PlasmaPy would differ
+  substantially from the version numbers of prior releases.
+
+Versioning Specification
+------------------------
+
+PlasmaPy final releases will use `calendar based versioning`_ of the
+form ``YY.0M.MICRO``, where:
+
+* The major release number ``YY`` is the year minus 2000,
```
Comment (namurphy) [inline PLEP-0008.rst:73]: Would this be better with a two-digit year or a four-digit year?

Comment (StanczakDominik) [reply PLEP-0008.rst:73]: 4-digit YYYY, for the same reasons as https://github.com/dask/community/issues/100#issuecomment-704499489 :)

File: PLEP-0008.rst [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+backwards incompatible changes to the API, ``MINOR`` is incremented
+when there are backwards compatible changes, and ``PATCH`` is
+incremented when the API does not change.  The main advantage of
+semantic versioning is that it is easy to tell when there are
+backwards incompatible changes. Because a community software project
+will have features at different stages of development, it is likely
+that there will be some backward incompatible changes during every
+release. Hence, date-based versioning is more appropriate to PlasmaPy
+than versioning based on whether or not backwards incompatible changes
+have been made.
```
Comment (StanczakDominik) [inline PLEP-0008.rst:106]: "Because a community software project will have features"... ? I think I get what you were going for here, but can you rephrase?

**Review by StanczakDominik — APPROVED** [review]

**Review by StanczakDominik — APPROVED** [review]

**Review by StanczakDominik — APPROVED** [review]

File: README.rst [hunk 3]
```diff
@@ -38,3 +38,6 @@ convenience, an abbreviated index of PLEPs is provided below.
 | 7      | `Structure of Top-Level Sub-Packages <./PLEP-0007.rst>`__                        | .. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.3774573.svg |
 |        |                                                                                  |    :target:  http://doi.org/10.5281/zenodo.3774573                 |
 +--------+----------------------------------------------------------------------------------+--------------------------------------------------------------------+
+| 8      | `Calendar Versioning <./PLEP-0008.rst>`__                                        |                                                                    |
+|        |                                                                                  |                                                                    |
```
Comment (rocco8773) [inline README.rst:42]: Do you plan to have a followup PR to put the doi in?  Or, will that be an update to this PR?

File: PLEP-0008.rst [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+.. note::
+
+    This PLEP supersedes |PLEP 5|, which is no longer in effect.
+
+Versioning Specification
+------------------------
+
+PlasmaPy final releases will use `calendar versioning`_ of the
+form ``YYYY.0M.MICRO``, where:
```
Comment (rocco8773) [inline PLEP-0008.rst:39]: Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

File: PLEP-0008.rst [hunk 5]
```diff
@@ -0,0 +1,166 @@
+PLEP-0008 – Calendar Versioning
+===============================
+
++-------------------+---------------------------------------------+
+| PLEP              | number                                      |
```
Comment (rocco8773) [inline PLEP-0008.rst:5]:
```suggestion
| PLEP              | 8                                           |
```

File: PLEP-0008.rst [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
++===================+=============================================+
+| title             | PlasmaPy Versioning and Releases            |
++-------------------+---------------------------------------------+
+| author(s)         | Nick Murphy                                 |
++-------------------+---------------------------------------------+
+| contact email     | namurphy@cfa.harvard.edu                    |
++-------------------+---------------------------------------------+
+| date created      | 2022-11-22                                  |
++-------------------+---------------------------------------------+
+| date last revised | 2022-11-29                                  |
```
Comment (rocco8773) [inline PLEP-0008.rst:15]:
```suggestion
| date last revised | 2022-12-08                                  |
```

[optional ... ?] just update to current date

File: PLEP-0008.rst [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
++-------------------+---------------------------------------------+
+| type              | process                                     |
++-------------------+---------------------------------------------+
+| status            | discussion                                  |
++-------------------+---------------------------------------------+
+| DOI               |                                             |
```
Comment (rocco8773) [inline PLEP-0008.rst:21]: update with reserved DOI

File: PLEP-0008.rst (hunk 4 above)
Comment (rocco8773) [inline PLEP-0008.rst:39]:
```suggestion
form ``YYYY.M.MICRO``, where:
```

Leading zeros will be removed in accordance with [PEP440](https://peps.python.org/pep-0440/#integer-normalization).  Basically these are normalized in accordance with `int()`.

File: PLEP-0008.rst [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    This PLEP supersedes |PLEP 5|, which is no longer in effect.
+
+Versioning Specification
+------------------------
+
+PlasmaPy final releases will use `calendar versioning`_ of the
+form ``YYYY.0M.MICRO``, where:
+
+* The major release number ``YYYY`` is the year,
+* The minor release number ``0M`` is the zero-padded month, and
```
Comment (rocco8773) [inline PLEP-0008.rst:42]:
```suggestion
* The minor release number ``M`` is positive integer representing the release month, and
```

See above PEP440 comment.  Feel to re-word to your own liking.

File: PLEP-0008.rst [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+* The bugfix release number ``MICRO`` will start from ``0`` and be
+  incremented by one for each bugfix release.
+
+For example, the first release in April 2023 would be ``2023.04.0``,
+with ``2023.04.1`` as the first bugfix release of the ``2023.04.*``
+series.
```
Comment (rocco8773) [inline PLEP-0008.rst:48]:
```suggestion
For example, the first release in April 2023 would be ``2023.4.0``,
with ``2023.4.1`` as the first bugfix release of the ``2023.4.*``
series.
```

Again, refer to the above PEP440 comments.

File: PLEP-0008.rst [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+* The age of a release can be gauged directly from the version number.
+
+* The extent of changes between releases of a package tends to be
+  roughly proportional to the time between releases, and can therefore
+  be gauged more accurately with date-based versioning.
+
+* The versioning scheme would be consistent with projects that use some
+  form of calendar versioning such as xarray, dask, pip, Ubuntu, and
+  PyCharm.
```
Comment (rocco8773) [inline PLEP-0008.rst:66]:
```suggestion
* The versioning scheme would be consistent with projects that use some
  form of calendar versioning such as ``xarray``, ``dask``, ``pip``,
  Ubuntu, and PyCharm.
```

[optional] ... but thank you for using the Oxford comma. 😺

File: PLEP-0008.rst [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+Bugfix releases shall contain no changes to the API.
+
+Tradeoff Analysis
+-----------------
+
+Advantages of adopting calendar versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The age of a release can be gauged directly from the version number.
```
Comment (rocco8773) [inline PLEP-0008.rst:58]: Not sure the best way to make the comment, but also easier to traceback the version when cited in a journal article.

File: PLEP-0008.rst [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  long-term support (LTS) release.
+
+* Versions of the form ``0.*.*`` (as currently in use) indicate that
+  breaking changes could occur for any release.
+
+Additional considerations for adopting calendar versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* A change to calendar versioning would need to be permanent in order to
+  preserve version monotonicity.
```
Comment (namurphy) [inline PLEP-0008.rst:111]:
```suggestion
```
There's something in PEP 440 about defining epochs, so this change need not be permanent.

File: PLEP-0008.rst [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+Additional considerations for adopting calendar versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Calendar versioning is compatible with :pep:`440` and
+  ``setuptools_scm``.
+
+* Using the full year in the version number will reduce the chance of
+  ambiguity. Similarly, using the zero-padded month will increase the
+  likelihood that the minor version number is interpreted as a month.
```
Comment (rocco8773) [inline PLEP-0008.rst:115]:
```suggestion
* Using the full year in the version number will reduce the chance of
  ambiguity. Similarly.
```

I'm just deleting this, but feel free to add back a comment about the month but without the "zero-padded" statement.

File: PLEP-0008.rst [hunk 14]
```diff
@@ -36,15 +36,15 @@ Versioning Specification
 ------------------------
 
 PlasmaPy final releases will use `calendar versioning`_ of the
-form ``YYYY.0M.MICRO``, where:
+form ``YYYY.MM.MICRO``, where:
 
 * The major release number ``YYYY`` is the year,
-* The minor release number ``0M`` is the zero-padded month, and
+* The minor release number ``MM`` is positive integer representing the release month, and
```
Comment (StanczakDominik) [inline PLEP-0008.rst:42]:
```suggestion
* The minor release number ``MM`` is a positive integer representing the release month, and
```

**Comment by StanczakDominik** [comment]
I'd just like to bring https://github.com/dask/community/issues/100#issuecomment-704468187 to your attention on the zero-padding discussion! The important points:

> 20.01.01 / 2020.01.01 are valid PEP 440 versions.

> Overall, the choice between 0Y/0M/0D vs YY/MM/DD is purely a aesthetic one -- either will work with basically every Python Packaging tool that follows the standards.

I'm weakly in favor of zero-padding, but don't mind much either way.

**Review by StanczakDominik — APPROVED** [review]

**Comment by namurphy** [comment]
I decided against discussing our LTS strategy since we don't have plans to do that in the near future, and I'd rather have a separate PLEP on that which we can do when we decide to do the LTS releases.

Other than that, I think I addressed all suggestions.  @rocco8773 — if this looks good to you, please feel free to merge.

File: PLEP-0005.rst [hunk 15]
```diff
@@ -16,12 +16,17 @@ PLEP-0005 – PlasmaPy Versioning and Releases
...
+| status            | superseded by |PLEP 8|                      |
 +-------------------+---------------------------------------------+
 | DOI               | `10.5281/zenodo.1451974                     |
 |                   | <https://doi.org/10.5281/zenodo.1451974>`__ |
 +-------------------+---------------------------------------------+
 
+.. important::
+
+    This PLEP has been superseded by |PLEP 8| and is therefore no longer
+    in effect.
```
Comment (namurphy) [inline PLEP-0005.rst:28]:
```suggestion
This PLEP has been superseded by |PLEP 8| and is therefore no longer
in effect.
```

File: PLEP-0000.rst [hunk 16]
```diff
@@ -2,81 +2,116 @@
...
+
+PLEPs that are no longer in effect
+==================================
+
++--------+-------------------------------------+------------------------+------------+---------------------------+
+| number | title                               | author(s)              | type       | DOI                       |
++========+=====================================+========================+============+===========================+
+| 5      | `PlasmaPy Versioning and Releases`_ | Nicholas A. Murphy and | process    | `10.5281/zenodo.1451974`_ |
+|        |                                     | Stuart J. Mumford      |            |                           |
++--------+--------------------------------------+------------------------+-----------+---------------------------+
```
Comment (namurphy) [inline PLEP-0000.rst:96]: Need to fix table!

File: PLEP-0005.rst [hunk 17]
```diff
@@ -257,3 +260,5 @@ Semantic versioning is a well-defined versioning scheme that provides
 users with useful information about whether or not there were any
 backward incompatible changes.  This scheme is well-suited to a core
 scientific software package that will require stability.
+
+.. |PLEP 8| replace:: https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0007.rst
```
Comment (rocco8773) [inline PLEP-0005.rst:264]:
```suggestion
.. |PLEP 7| replace:: https://github.com/PlasmaPy/PlasmaPy-PLEPs/blob/main/PLEP-0007.rst
```

File: README.rst (hunk 3 above)
Comment (rocco8773) [reply README.rst:42]: Ping... Do we have a DOI to put here now?

File: README.rst [hunk 18]
```diff
@@ -38,3 +38,6 @@ convenience, an abbreviated index of PLEPs is provided below.
 | 7      | `Structure of Top-Level Sub-Packages <./PLEP-0007.rst>`__                        | .. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.3774573.svg |
 |        |                                                                                  |    :target:  http://doi.org/10.5281/zenodo.3774573                 |
 +--------+----------------------------------------------------------------------------------+--------------------------------------------------------------------+
+| 8      | `Calendar Versioning <./PLEP-0008.rst>`__                                        |                                                                    |
```
Comment (rocco8773) [inline README.rst:41]: Do we have a DOI to put here now?

File: PLEP-0000.rst [hunk 19]
```diff
@@ -2,81 +2,116 @@
...
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 2      | `PlasmaPy Governance`_                                       | Nicholas A. Murphy   | process    | `10.5281/zenodo.1435980`_ |
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 4      | `Licensing of PlasmaPy Repositories`_                        | Nicholas A. Murphy   | process    | `10.5281/zenodo.1435990`_ |
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 6      | `A New General-Purpose Plasma Object`_                       | Andrew J. Leonard    | standard   | `10.5281/zenodo.1460976`_ |
++--------+--------------------------------------------------------------+----------------------+------------+---------------------------+
+| 7      | `Structure of Top-Level Sub-Packages`_                       | Erik T. Everson,     | standard   |                           |
```
Comment (rocco8773) [inline PLEP-0000.rst:53]: This is missing its DOI.

**Review by rocco8773 — APPROVED** [review]

File: PLEP-0000.rst (hunk 19 above)
Comment (namurphy) [reply PLEP-0000.rst:53]: This hasn't been uploaded to Zenodo, so it does not yet have a DOI.

Comment (namurphy) [reply PLEP-0000.rst:53]: I lied. It does have one, which I'll add.

 - [ ] Do ↑
